pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+/*
+NOTE: this needs to be tested in e2e since it uses iptables for everything.
+*/
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util"
+	utilexec "k8s.io/kubernetes/pkg/util/exec"
+	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	"k8s.io/kubernetes/pkg/util/slice"
+)
+
+// NOTE: It will be tricky to determine the exact version needed.
+// Some features will of course be backported in various distros and this could get pretty hairy.
+// However iptables-1.4.0 was released 2007-Dec-22 and appears to have every feature we use,
+// so this seems prefectly reasonable for now.
+const (
+	IPTABLES_MIN_MAJOR = 1
+	IPTABLES_MIN_MINOR = 4
+	IPTABLES_MIN_PATCH = 0
+)
+
+// semVerGreaterOrEqual returns true if the semantic version (major,minor,patch)
+// is greater than or equal to (minMajor, minMinor, minPatch)
+func semVerGreaterOrEqual(major, minor, patch, minMajor, minMinor, minPatch int) bool {
+	if major != minMajor {
+		return major > minMajor
+	}
+	if minor != minMinor {
+		return minor > minMinor
+	}
+	return patch >= minPatch
+}
+
+// ShouldUseIptablesProxier returns true if we should use the iptables Proxier instead of
+// the userspace Proxier.
+// This is determined by the iptables version. It may return an erorr if it fails to get the
+// itpables version without error, in which case it will also return false.
+func ShouldUseIptablesProxier() (bool, error) {
+	exec := utilexec.New()
+	major, minor, patch, err := utiliptables.GetIptablesVersion(exec)
+	if err != nil {
+		return false, err
+	}
+	return semVerGreaterOrEqual(major, minor, patch,
+		IPTABLES_MIN_MAJOR, IPTABLES_MIN_MINOR, IPTABLES_MIN_PATCH), err
+}
+
+type portal struct {
+	ip   net.IP
+	port int
+}
+
+// A key for the portMap
+type portMapKey struct {
```

Comment:
Nope, not used anywhere. Previously used for the EndpointTracking... Bad
copy/paste when grabbing the other utility methods.
Removing.

On Mon, Aug 10, 2015 at 7:04 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36695236>
> :
>
> > +	exec := utilexec.New()
> > +	major, minor, patch, err := utiliptables.GetIptablesVersion(exec)
> > +	if err != nil {
> > +		return false, err
> > +	}
> > +	return semVerGreaterOrEqual(major, minor, patch,
> > +		IPTABLES_MIN_MAJOR, IPTABLES_MIN_MINOR, IPTABLES_MIN_PATCH), err
> > +}
> > +
> > +type portal struct {
> > +	ip   net.IP
> > +	port int
> > +}
> > +
> > +// A key for the portMap
> > +type portMapKey struct {
>
> not used anywhere?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36695236>
> .
>